### PR TITLE
feat: Add Explanations to service APIs

### PR DESCRIPTION
**Pull request by vinnysenthil**
- Lifts the GAPIC types `ExplanationMetadata` and `ExplanationParameters` to the `aiplatform` namespace
- Adds the ability to pass the two fields in [`ExplanationSpec`](https://github.com/googleapis/python-aiplatform/blob/cdeb0ec30c334ff2b5d5e06bc976e824d6e18c04/google/cloud/aiplatform_v1beta1/types/explanation.py#L247-L263) to the following methods
  - `Model.batch_predict()`
  - `Model.deploy()`
  - `Model.upload()`
  - `Endpoint.deploy()`
  - `BatchPredictionJob.create()`
- Add the method `Endpoint.explain()` to get predictions with explanations on deployed models
- Unit tests for all the above changes

Fixes [b/174502675](http://b/174502675) and part 1 of [b/172828587](http://b/172828587) 🦕

### CODE REVIEW

**Review by ivanmkc — APPROVED** [review]
I haven't used explanations before but otherwise LGTM

**Review by sasha-gitg — CHANGES_REQUESTED** [review]
Added a few comments. Thanks!

File: google/cloud/aiplatform/__init__.py
Comment (vinnysenthil) [reply google/cloud/aiplatform/__init__.py:41]: Sounds good, will elevate all Explanation protos to `aiplatform.explain`

File: google/cloud/aiplatform/jobs.py
Comment (vinnysenthil) [reply google/cloud/aiplatform/jobs.py:465]: Nice catch, I'll expose `generate_explanation` as a parameter

File: google/cloud/aiplatform/__init__.py
Comment (vinnysenthil) [reply google/cloud/aiplatform/__init__.py:41]: Actually only including protos that are used as inputs, and only raising `ExplanationMetadata` subclasses that are two levels deep. Lmk if you think we should include outputs too.

File: google/cloud/aiplatform/jobs.py
Comment (vinnysenthil) [reply google/cloud/aiplatform/jobs.py:319]: Thanks for catching this, also allowed `explanation_metadata` to be passed independently of `explanation_parameters` since all these fields are for optional overriding.

**Review by sasha-gitg — APPROVED** [review]
LGTM. Left a two more comments.

File: google/cloud/aiplatform/jobs.py
Comment (vinnysenthil) [reply google/cloud/aiplatform/jobs.py:232]: Yup, I'll change this to False. A note: we only set the `generate_explanation` field in the request if the user sets the flag to `True`

File: google/cloud/aiplatform/models.py
Comment (vinnysenthil) [reply google/cloud/aiplatform/models.py:754]: I've confirmed the service throws if only one part is provided, let's populate the spec if either is provided and let the service throw an error. We inform the user in the docstrings that they need to be passed together.

**Comment by vinnysenthil** [comment]
All requested changes from MB SDK Dev addressed. Blocking merge until receiving LGTM from XAI Eng.

File: google/cloud/aiplatform/explain/__init__.py
Comment (besimav) [inline google/cloud/aiplatform/explain/__init__.py:41]: a bit orthogonal question: why don't we have Modality class?

**Review by besimav — COMMENTED** [review]
LGTM overall.

File: google/cloud/aiplatform/explain/__init__.py
Comment (vinnysenthil) [reply google/cloud/aiplatform/explain/__init__.py:41]: It appears that `modality` is a string field in `InputMetadata` [according to the protos](http://google3/google/cloud/aiplatform/master/explanation_metadata.proto?l=289-294&rcl=358240862). Should it be an enum?

Comment (besimav) [reply google/cloud/aiplatform/explain/__init__.py:41]: I was looking at [this proto](https://source.corp.google.com/piper///depot/google3/cloud/ai/platform/db/proto/explanation.proto;l=250). Do you know why it is an Enum in the DB one and not in the other by any chance? Anyhow, lgtm.

Comment (vinnysenthil) [reply google/cloud/aiplatform/explain/__init__.py:41]: Interesting, I raised this with jianan@ and he explained they lean towards string over enums in the API protos to avoid having to make frequent changes to those files while the DB protos are more strongly typed. If it improves usability, we could manually add the enum in the SDK but it would have to be kept in sync.

**Review by nanchenchen — APPROVED** [review]
Thanks for working on this! It looks good to me 👍 
I only have a minor comment. Should be good to go after addressing it.

File: google/cloud/aiplatform/models.py
Comment (vinnysenthil) [reply google/cloud/aiplatform/models.py:759]: The thought behind this was allowing the service to throw the exception about them being passed together instead of creating two sets of validation. I confirmed that an error response is returned service-side right away if only one is passed.

Comment (vinnysenthil) [reply google/cloud/aiplatform/models.py:759]: We inform the user in the docstring. Do you think we should add validation in the library as well?
https://github.com/googleapis/python-aiplatform/blob/b5000c2d0f27c4a746bb07c499918d54092a7458/google/cloud/aiplatform/models.py#L480-L481

Comment (nanchenchen) [reply google/cloud/aiplatform/models.py:759]: I think telling users what will fail as early as possible will be better. We don't necessary need full validation (though it is good to have if it is low cost), but having some basic sanity check can help prevent user frustration (especially since all the operations here are async and have some wait time). 

Also, I noticed that there other similar logics are using `and` (e.g., Ln 1287), so I thought it was unintentional :)

Comment (vinnysenthil) [reply google/cloud/aiplatform/models.py]: Good point, I've added the local validation for ensuring they are passed together.

**Pull request merged**
